File: topics/examples/pulsar-producer-example.py
#!/usr/bin/env python3
"""
Pulsar Producer示例

功能：
- 创建Producer并发送消息
- 支持同步和异步发送
- 支持批量发送
"""

import pulsar
from pulsar import MessageId

# 配置
SERVICE_URL = "pulsar://localhost:6650"
TOPIC = "persistent://public/default/my-topic"

def main():
    # 创建Pulsar客户端
    client = pulsar.Client(SERVICE_URL)
    
    try:
        # 创建Producer
        producer = client.create_producer(
            topic=TOPIC,
            producer_name="my-producer",
            batching_enabled=True,
            batching_max_messages=100,
            batching_max_publish_delay_ms=10
        )
        
        # 同步发送消息
        msg_id = producer.send("Hello Pulsar".encode('utf-8'))
        print(f"Message sent: {msg_id}")
        
        # 异步发送消息
        producer.send_async(
            "Hello Pulsar Async".encode('utf-8'),
            callback=lambda res, msg_id: print(f"Message sent async: {msg_id}")
        )
        
        # 发送带属性的消息
        producer.send(
            "Hello Pulsar with Properties".encode('utf-8'),
            properties={'property1': 'value1'},
            partition_key='my-key'
        )
        
        # 发送多条消息
        for i in range(10):
            producer.send(f"Message {i}".encode('utf-8'))
        
        # 关闭Producer
        producer.close()
    finally:
        # 关闭客户端
        client.close()

if __name__ == "__main__":
    main()

File: topics/examples/pulsar-producer-example.java
import org.apache.pulsar.client.api.*;

/**
 * Pulsar Producer示例
 * 
 * 功能：
 * - 创建Producer并发送消息
 * - 支持同步和异步发送
 * - 支持批量发送
 * - 支持事务发送
 */
public class PulsarProducerExample {
    private static final String SERVICE_URL = "pulsar://localhost:6650";
    private static final String TOPIC = "persistent://public/default/my-topic";
    
    public static void main(String[] args) throws Exception {
        // 创建Pulsar客户端
        PulsarClient client = PulsarClient.builder()
            .serviceUrl(SERVICE_URL)
            .build();
        
        try {
            // 创建Producer
            Producer<byte[]> producer = client.newProducer()
                .topic(TOPIC)
                .producerName("my-producer")
                .enableBatching(true)
                .batchingMaxMessages(100)
                .batchingMaxPublishDelay(10, TimeUnit.MILLISECONDS)
                .create();
            
            // 同步发送消息
            MessageId msgId = producer.send("Hello Pulsar".getBytes());
            System.out.println("Message sent: " + msgId);
            
            // 异步发送消息
            producer.sendAsync("Hello Pulsar Async".getBytes())
                .thenAccept(msgId -> {
                    System.out.println("Message sent async: " + msgId);
                });
            
            // 发送带属性的消息
            producer.newMessage()
                .key("my-key")
                .value("Hello Pulsar with Key".getBytes())
                .property("property1", "value1")
                .send();
            
            // 事务发送（需要启用事务）
            /*
            Transaction txn = client.newTransaction()
                .withTransactionTimeout(30, TimeUnit.SECONDS)
                .build()
                .get();
            
            producer.newMessage(txn)
                .value("Transaction Message".getBytes())
                .send();
            
            txn.commit().get();
            */
            
            // 关闭Producer
            producer.close();
        } finally {
            // 关闭客户端
            client.close();
        }
    }
}
